### PR TITLE
Double backward reduce mean

### PR DESCRIPTION
**Pull request by ceci3**
**Add reduce_mean_grad_grad op for backward of backward calculation**

### CODE REVIEW

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc [hunk 1]
```diff
@@ -13,8 +13,72 @@
...
+    op->SetType("reduce_mean_grad");
+    op->SetInput("X", Input("X"));
+    op->SetInput(framework::GradVarName("Out"), OutputGrad("Out"));
+    op->SetAttrMap(Attrs());
+    op->SetOutput(framework::GradVarName("X"), InputGrad("X"));
+    return op;
+  }
+};
+
+class ReduceMeanGradGradMaker: public framework::GradOpDescMakerBase{
```
Comment (qingqing01) [inline paddle/fluid/operators/reduce_ops/reduce_mean_op.cc]: GradGrad -> DoubleGrad

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc [hunk 2]
```diff
@@ -13,8 +13,74 @@
...
 #include "paddle/fluid/operators/reduce_ops/reduce_mean_op.h"
+#include <memory>
+#include <string>
+#include <vector>
 
-REGISTER_REDUCE_OP(reduce_mean);
+namespace paddle {
+namespace operators {
+
+class ReduceMeanOpGradDescMaker : public framework::SingleGradOpDescMaker {
```
Comment (qingqing01) [inline paddle/fluid/operators/reduce_ops/reduce_mean_op.cc:27]: Add comments for why rewrite the `ReduceMeanOpGradDescMaker`

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc [hunk 3]
```diff
@@ -13,8 +13,74 @@
...
+    op->SetType("reduce_mean_grad");
+    op->SetInput("X", Input("X"));
+    op->SetInput(framework::GradVarName("Out"), OutputGrad("Out"));
+    op->SetAttrMap(Attrs());
+    op->SetOutput(framework::GradVarName("X"), InputGrad("X"));
+    return op;
+  }
+};
+
+class ReduceMeanGradGradMaker : public framework::GradOpDescMakerBase {
```
Comment (qingqing01) [inline paddle/fluid/operators/reduce_ops/reduce_mean_op.cc]: ReduceMeanGradGradMaker -> ReduceMeanDoubleGradMaker

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc [hunk 4]
```diff
@@ -13,8 +13,74 @@
...
+  using framework::GradOpDescMakerBase::GradOpDescMakerBase;
+
+  std::vector<std::unique_ptr<framework::OpDesc>> operator()() const override {
+    std::vector<std::unique_ptr<framework::OpDesc>> ops;
+    auto x_grads = InputGrad("X");
+    if (!x_grads.empty()) {
+      auto* x_grad_op = new framework::OpDesc();
+      x_grad_op->SetType("scale");
+      // input: ddx
+      x_grad_op->SetInput("X", OutputGrad(framework::GradVarName("X")));
```
Comment (heavengate) [inline paddle/fluid/operators/reduce_ops/reduce_mean_op.cc]: 这里和61行都有OutputGrad(framework::GradVarName("X"))，抽出来做个变量，这两行填变量吧，代码感觉清晰些

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc (hunk 2 above)
Comment (heavengate) [reply paddle/fluid/operators/reduce_ops/reduce_mean_op.cc:27]: Done

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc (hunk 4 above)
Comment (heavengate) [reply paddle/fluid/operators/reduce_ops/reduce_mean_op.cc]: Done

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc (hunk 1 above)
Comment (heavengate) [reply paddle/fluid/operators/reduce_ops/reduce_mean_op.cc]: Done

File: paddle/fluid/operators/reduce_ops/reduce_mean_op.cc (hunk 3 above)
Comment (heavengate) [reply paddle/fluid/operators/reduce_ops/reduce_mean_op.cc]: Done

**Review by qingqing01 — DISMISSED** [review]

**Review by heavengate — APPROVED** [review]
LGTM

**Comment by heavengate** [comment]
```+ '[' paddle/fluid/op_use_default_grad_op_maker.spec == CMakeLists.txt ']'
[05:56:08]	+ '[' paddle/fluid/op_use_default_grad_op_maker.spec == python/paddle/fluid/__init__.py ']'
[05:56:08]	++ curl -H 'Authorization: token 5ae37427a95d9e51fb0ea3c71121b75d5c2ebefa' 'https://api.github.com/repos/PaddlePaddle/Paddle/pulls/17372/reviews?per_page=10000'
[05:56:08]	++ python /paddle/tools/check_pr_approval.py 1 21351065 3048612 46782768 30176695 12538138 6836917 32832641
[05:56:08]	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[05:56:08]	                                 Dload  Upload   Total   Spent    Left  Speed
[05:56:08]	
[05:56:08]	  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[05:56:10]	  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[05:56:10]	  1 12676    1   138    0     0     73      0  0:02:53  0:00:01  0:02:52    73
[05:56:10]	100 12676  100 12676    0     0   6075      0  0:00:02  0:00:02 --:--:--  6076
[05:56:10]	+ APPROVALS=FALSE
[05:56:10]	+ echo 'current pr 17372 got approvals: FALSE'
```
please approve for op_use_default_grad_op_maker.spec change @tensor-tang

**Review by tensor-tang — APPROVED** [review]
LGTM

**Pull request merged**